src/signature.c

```diff
@@ -24,198 +24,386 @@
+	 * in which to search for the signer's cert. */
+	x509_stack = sk_X509_new_null();
+	if (!x509_stack) {
+		fprintf(stderr, "Failed to initialize X509 certificate stack\n");
+		goto error;
+	}
+	sk_X509_push(x509_stack, cert);
+
+	/* Verify the signature, outdata can be NULL because we don't use it */
+	ret = PKCS7_verify(p7, x509_stack, store, indata, NULL, 0);
+	if (ret == 1) {
+		printf("VERIFY SUCCESS!\n");
+		return true;
+	}
+error:
+	if (signature) {
+		free(signature);
+	}
+	if (data) {
+		free(data);
+	}
+	ERR_print_errors_fp(stderr);
+	fprintf(stderr, "Signature check failed\n");
+	return false;
 }
 
-void signature_delete(const char *data_filename)
+/* This function makes sure the certificate is still valid by not having any
+ * compromised certificates in the chain.
+ * If there is no Certificate Revocation List (CRL) it may be that the private
+ * keys have not been compromised or the CRL has not been generated by the
+ * Certificate Authority (CA)
+ *
+ * returns: true if certificate is valid, false otherwise */
+static bool validate_certificate(void)
 {
-	char *sig_filename;
+	FILE *fp = NULL;
+	X509_LOOKUP *lookup = NULL;
+	X509_STORE_CTX *verify_ctx = NULL;
+
+	/* CRL and Chains are not required for the current setup, but we may
+	 * implement them in the future */
+	if (!crl) {
+		printf("No certificate revocation list provided\n");
+	}
+	if (!chain) {
+		printf("No certificate chain provided\n");
+	}
 
-	string_or_die(&sig_filename, "%s.signed", data_filename);
+	/* create the cert store and set the verify callback */
+	if (!(store = X509_STORE_new())) {
+		goto error;
+	}
 
-	unlink(sig_filename);
+	X509_STORE_set_verify_cb_func(store, verify_callback);
 
-	free(sig_filename);
-}
+	/* Add the certificates to be verified to the store */
+	if (!(lookup = X509_STORE_add_lookup(store, X509_LOOKUP_file()))) {
+		goto error;
+	}
 
-#else // IMPL == IMPL_FAKE
+	/*  Load the our Root cert, which can be in either DER or PEM format */
+	if (!X509_load_cert_file(lookup, CERTNAME, X509_FILETYPE_PEM)) {
+		goto error;
+	}
 
-bool signature_initialize(const char UNUSED_PARAM *ca_cert_filename)
-{
+	if (crl) {
+		if (!(lookup = X509_STORE_add_lookup(store, X509_LOOKUP_file())) ||
+		   (X509_load_crl_file(lookup, crl, X509_FILETYPE_PEM) != 1)) {
+			goto error;
+		}
+		/* set the flags of the store so that CLRs are consulted */
+		X509_STORE_set_flags(store, X509_V_FLAG_CRL_CHECK | X509_V_FLAG_CRL_CHECK_ALL);
+	}
+
+	/* create a verification context and initialize it */
+	if (!(verify_ctx = X509_STORE_CTX_new())) {
+		goto error;
+	}
+
+	if (X509_STORE_CTX_init(verify_ctx, store, cert, NULL) != 1) {
+		goto error;
+	}
+
+	/* Specify which cert to validate in the verify context.
+	 * This is required because we may add multiple certs to the X509 store,
+	 * but we want to validate a specific one out of the group/chain. */
+	X509_STORE_CTX_set_cert(verify_ctx, cert);
+
+	/* verify the certificate */
```

Comment:
I don't see another comment?  I'm leaving validate_certificate() as implemented for now.  The call is expected to succeed, and if it does not succeed that is an informative sign even today.